ethexe/consensus/src/utils.rs

```diff
@@ -0,0 +1,354 @@
+use std::collections::BTreeMap;
+
+/// Represents a request for validating a batch of block commitments.
+/// This structure is used to verify the integrity and validity of multiple block commitments
+/// and their associated code commitments.
+#[derive(Debug, Clone, Encode, Decode, PartialEq, Eq)]
+pub struct BatchCommitmentValidationRequest {
+    /// List of block commitment validation requests
+    pub blocks: Vec<BlockCommitmentValidationRequest>,
+    /// List of code commitments to be validated
+    pub codes: Vec<CodeCommitment>,
+}
+
+impl From<&BatchCommitment> for BatchCommitmentValidationRequest {
+    fn from(batch: &BatchCommitment) -> Self {
+        BatchCommitmentValidationRequest {
+            blocks: batch
+                .block_commitments
+                .iter()
+                .map(BlockCommitmentValidationRequest::from)
+                .collect(),
+            codes: batch.code_commitments.clone(),
+        }
+    }
+}
+
+impl ToDigest for BatchCommitmentValidationRequest {
+    fn update_hasher(&self, hasher: &mut ethexe_signer::sha3::Keccak256) {
+        hasher.update(self.codes.to_digest().as_ref());
+        hasher.update(self.blocks.to_digest().as_ref());
+    }
+}
+
+/// A request for validating a single block commitment.
+/// Contains all necessary information to verify the integrity of a block's commitment.
+///
+/// NOTE: [`BlockCommitmentValidationRequest`] digest is always equal to the corresponding
+/// [`BlockCommitment`] digest.
+#[derive(Debug, Clone, Encode, Decode, PartialEq, Eq)]
+pub struct BlockCommitmentValidationRequest {
+    /// Hash of the block being validated
+    pub block_hash: H256,
+    /// Timestamp of the block
+    pub block_timestamp: u64,
+    /// Hash of the previous non-empty block
+    pub previous_not_empty_block: H256,
+    /// Hash of the predecessor block
+    pub predecessor_block: H256,
+    /// Digest of the block's state transitions
+    pub transitions_digest: Digest,
+}
+
+impl From<&BlockCommitment> for BlockCommitmentValidationRequest {
+    fn from(commitment: &BlockCommitment) -> Self {
+        BlockCommitmentValidationRequest {
+            block_hash: commitment.hash,
+            block_timestamp: commitment.timestamp,
+            previous_not_empty_block: commitment.previous_committed_block,
+            predecessor_block: commitment.predecessor_block,
+            transitions_digest: commitment.transitions.to_digest(),
+        }
+    }
+}
+
+impl ToDigest for BlockCommitmentValidationRequest {
+    fn update_hasher(&self, hasher: &mut ethexe_signer::sha3::Keccak256) {
+        hasher.update(self.block_hash.as_bytes());
+        hasher
+            .update(ethexe_common::u64_into_uint48_be_bytes_lossy(self.block_timestamp).as_slice());
+        hasher.update(self.previous_not_empty_block.as_bytes());
+        hasher.update(self.predecessor_block.as_bytes());
+        hasher.update(self.transitions_digest.as_ref());
+    }
```

Comment:
```suggestion
        // this would help not to forget to update this impl on changes to inner data 
        let Self { a, b, c } = self;
        
        hasher.update(self.block_hash.as_bytes());
        hasher
            .update(ethexe_common::u64_into_uint48_be_bytes_lossy(self.block_timestamp).as_slice());
        hasher.update(self.previous_not_empty_block.as_bytes());
        hasher.update(self.predecessor_block.as_bytes());
        hasher.update(self.transitions_digest.as_ref());
    }
```

Reply:
Done